README.md
# unexpected-http

[Unexpected](http://github.com/unexpectedjs/unexpected) plugin for testing HTTP servers. Uses same syntax as [unexpected-express](https://github.com/unexpectedjs/unexpected-express) whenever possible.

[![NPM version](https://badge.fury.io/js/unexpected-http.svg)](http://badge.fury.io/js/unexpected-http)
[![Build Status](https://travis-ci.org/unexpectedjs/unexpected-http.svg?branch=master)](https://travis-ci.org/unexpectedjs/unexpected-http)
[![Coverage Status](https://coveralls.io/repos/unexpectedjs/unexpected-http/badge.svg)](https://coveralls.io/r/unexpectedjs/unexpected-http)
[![Dependency Status](https://david-dm.org/unexpectedjs/unexpected-http.svg)](https://david-dm.org/unexpectedjs/unexpectetd-http)

Read [the documentation](http://unexpected.js.org/unexpected-http).


CHANGELOG.md
### v8.1.0 (2022-04-28)

- [Add unexpectedHttp.js to .prettierignore](https://github.com/unexpectedjs/unexpected-http/commit/a8c8edc3f030f322868501752f0d72a47526d2d9) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [eslint --fix .](https://github.com/unexpectedjs/unexpected-http/commit/3a00125a543994d5656ebc33064eeab96b6d5291) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Unsupport node.js 12](https://github.com/unexpectedjs/unexpected-http/commit/de9ae652ad08bfcbd335cc2e42f2c131b95f2f14) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Fiddle with the rollup setup](https://github.com/unexpectedjs/unexpected-http/commit/80577d9901b2fd661b0ec17b48a4f43ca1693918) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Add unexpectedHttp.js to .eslintignore](https://github.com/unexpectedjs/unexpected-http/commit/0d59267af0965037f2e8139ab9e37c73485d64e9) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [+5 more](https://github.com/unexpectedjs/unexpected-http/compare/v8.1.0...v8.1.0)

### v8.1.0 (2022-04-28)

#### Pull requests

- [#116](https://github.com/unexpectedjs/unexpected-http/pull/116) Upgrade puppeteer to version 13.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#115](https://github.com/unexpectedjs/unexpected-http/pull/115) Upgrade eslint-plugin-mocha to version 10.0.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#112](https://github.com/unexpectedjs/unexpected-http/pull/112) Upgrade puppeteer to version 12.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#111](https://github.com/unexpectedjs/unexpected-http/pull/111) Upgrade prettier to version 2.5.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#108](https://github.com/unexpectedjs/unexpected-http/pull/108) Upgrade puppeteer to version 11.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#103](https://github.com/unexpectedjs/unexpected-http/pull/103) Upgrade prettier to version 2.4.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#105](https://github.com/unexpectedjs/unexpected-http/pull/105) Upgrade eslint-plugin-promise to version 5.1.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#96](https://github.com/unexpectedjs/unexpected-http/pull/96) Upgrade puppeteer to version 10.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#95](https://github.com/unexpectedjs/unexpected-http/pull/95) Upgrade eslint-plugin-mocha to version 9.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#91](https://github.com/unexpectedjs/unexpected-http/pull/91) Upgrade puppeteer to version 9.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#87](https://github.com/unexpectedjs/unexpected-http/pull/87) Upgrade puppeteer to version 8.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#84](https://github.com/unexpectedjs/unexpected-http/pull/84) Upgrade magicpen-prism to version 5.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#83](https://github.com/unexpectedjs/unexpected-http/pull/83) Upgrade puppeteer to version 7.0.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#82](https://github.com/unexpectedjs/unexpected-http/pull/82) Upgrade puppeteer to version 6.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#81](https://github.com/unexpectedjs/unexpected-http/pull/81) Upgrade karma to version 6.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#80](https://github.com/unexpectedjs/unexpected-http/pull/80) Upgrade eslint-config-prettier to version 7.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#78](https://github.com/unexpectedjs/unexpected-http/pull/78) Upgrade eslint-plugin-standard to version 5.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))

#### Commits to master

- [Update unexpected the dev dep to ^13.0.0, allow 13 for the peer dep](https://github.com/unexpectedjs/unexpected-http/commit/a8c0c1ba98d37d61f22bb7fc3065f205364458aa) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [Whoops, make eslint-plugin-n a dev dep](https://github.com/unexpectedjs/unexpected-http/commit/53d1183cc112fab41b83f464d370dddae9e06630) ([Andreas Lind](mailto:andreas.lind@workday.com))
- [npm i eslint-plugin-n eslint-config-standard@latest && npm uninstall eslint-plugin-node eslint-plugin-standard](https://github.com/unexpectedjs/unexpected-http/commit/87032198183ba227cc86789d2553d7138ff381d1) ([Andreas Lind](mailto:andreas.lind@workday.com))
- [Update eslint-plugin-promise to version 6.0.0](https://github.com/unexpectedjs/unexpected-http/commit/09acd3738c005b968cda9391183638132dcecec8) ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [Unsupport node.js 10](https://github.com/unexpectedjs/unexpected-http/commit/25eeb7f9e5031033036f33e9626e6d793ca8f258) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))
- [+8 more](https://github.com/unexpectedjs/unexpected-http/compare/v8.0.1...v8.1.0)

### v8.0.1 (2020-11-22)

- [Update unexpected-messy to ^10.0.0](https://github.com/unexpectedjs/unexpected-http/commit/904b716974099b1aabf616b278199985d4388ce5) ([Andreas Lind](mailto:andreaslindpetersen@gmail.com))

### v8.0.0 (2020-11-22)

#### Pull requests

- [#73](https://github.com/unexpectedjs/unexpected-http/pull/73) Upgrade eslint-config-standard to version 16.0.1 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#71](https://github.com/unexpectedjs/unexpected-http/pull/71) Upgrade eslint-config-standard to version 15.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))
- [#69](https://github.com/unexpectedjs/unexpected-http/pull/69) Upgrade browserify to version 17.0.0 ([depfu[bot]](mailto:23717796+depfu[bot]@users.noreply.github.com))

test/unexpectedHttp.js
const unexpected = require('unexpected');
const http = require('http');
const semver = require('semver');
const stream = require('stream');

function createGetAddrInfoError(host, port) {
  let getaddrinfoError;
  // Different versions of node have shuffled around the properties of error instances:
  const nodeJsVersion = process.version.replace(/^v/, '');
  if (nodeJsVersion === '0.10.29') {
    getaddrinfoError = new Error('getaddrinfo EADDRINFO');
    getaddrinfoError.code = getaddrinfoError.errno = 'EADDRINFO';
  } else if (semver.satisfies(nodeJsVersion, '>=0.12.0')) {
    const message =
      'getaddrinfo ENOTFOUND www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com';
    if (semver.satisfies(nodeJsVersion, '>=9.7.0 <10')) {
      // https://github.com/nodejs/node/issues/19716
      getaddrinfoError = new Error();
      getaddrinfoError.message = message;
    } else {
      getaddrinfoError = new Error(message);
    }
    if (semver.satisfies(nodeJsVersion, '>=2.0.0 <12')) {
      getaddrinfoError.message += ` ${host}:${port}`;
      getaddrinfoError.host = host;
      getaddrinfoError.port = port;
    }
    getaddrinfoError.code = getaddrinfoError.errno = 'ENOTFOUND';
    if (semver.satisfies(nodeJsVersion, '>=13')) {
      getaddrinfoError.errno = -3008;
    }
    getaddrinfoError.hostname = 'www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com';
  } else {
    getaddrinfoError = new Error('getaddrinfo ENOTFOUND');
    getaddrinfoError.code = getaddrinfoError.errno = 'ENOTFOUND';
  }
  getaddrinfoError.syscall = 'getaddrinfo';
  return getaddrinfoError;
}

describe('unexpected-http', function () {
  const expect = unexpected
    .clone()
    .use(require('../lib/unexpectedHttp'))
    .addAssertion(
      '<any> when delayed a little bit <assertion?>',
      function (expect) {
        return expect.promise(function (run) {
          setTimeout(
            run(function () {
              return expect.shift();
            }),
            1
          );
        });
      }
    );

  expect.output.preferredWidth = 80;

  expect.output.installPlugin(require('magicpen-prism'));

  it('should do a basic request @integration', function () {
    return expect(
      'GET http://www.gofish.dk/',
      'to yield HTTP response satisfying',
      {
        headers: {
          'Content-Type': /text\/html/,
        },
      }
    );
  });

  it('should fail with a diff @integration', function () {
    return expect(
      expect('GET http://www.gofish.dk/', 'to yield HTTP response satisfying', {
        headers: {
          'Content-Type': /text\/plain/,
        },
      }),
      'when rejected to have message',
      /Content-Type: text\/html.*\/\/ should match \/text\\\/plain/
    );
  });

  describe('with the expected response object containing an error property @integration', function () {
    it('should expect an error #2', function () {
      return expect(
        'GET http://www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com/',
        'to yield HTTP response satisfying',
        {
          error: createGetAddrInfoError(
            'www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com',
            80
          ),
        }
      );
    });
  });

  describe('with the response property given as an error instance @integration', function () {
    it('should expect an error', function () {
      return expect(
        'GET http://www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com/',
        'to yield HTTP response satisfying',
        createGetAddrInfoError('www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com', 80)
      );
    });

    it('should fail with a diff if the request fails with an error that does not equal the expected error', function () {
      return expect(
        expect(
          'GET http://www.veqwjioevjqwoevijqwokevijqwioevjkqwioevq.com/',
          'to yield HTTP response satisfying',
          new Error('foobar')
        ),
        'when rejected to have message',
        expect.it(function (message) {
          // The error varies a lot depending on the node.js version:
          expect(
            message.replace(/Error\(\{[\s\S]*\}\)$/, 'Error(...)'),
            'to equal',
            "expected 'GET http://www.veqwjioevjqwoevijqwokevijqwioevjkqwioevq.com/'\n" +
              "to yield HTTP response satisfying Error('foobar')\n" +
              '\n' +
              'Error(...)'
          );
        })
      );
    });
  });

  it('should expect an error if the response is given as an error @integration', function () {
    return expect(
      'GET http://www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com/',
      'to yield HTTP response satisfying',
      createGetAddrInfoError('www.icwqjecoiqwjecoiwqjecoiwqjceoiwq.com', 80)
    );
  });

  it('should reject with an actual UnexpectedError mentioning the error code when an unexpected socket error is encountered', function () {
    return expect(
      expect(
        'GET http://www.veqwjioevjqwoevijqwokevijqwioevjkqwioevq.com/',
        'to yield HTTP response satisfying',
        {}
      ),
      'when rejected to have message',
      /getaddrinfo/
    );
  });

  describe('with a local test server', function () {
    let handleRequest, serverHostname, serverAddress, serverUrl, server;

    beforeEach(function () {
      handleRequest = undefined;
      server = http.createServer(function (req, res) {
        res.sendDate = false;
        handleRequest(req, res);
      });
      server.listen(0);
      serverAddress = server.address();
      serverHostname =
        serverAddress.address === '::' ? 'localhost' : serverAddress.address;
      serverUrl = `http://${serverHostname}:${serverAddress.port}/`;
    });

    afterEach(function () {
      server.close();
    });

    it('should provide a context object', function () {
      handleRequest = function (req, res) {
        res.writeHead(503);
        res.end();
      };
      return expect(serverUrl, 'to yield HTTP response satisfying', 503).then(
        function (context) {
          expect(context, 'to satisfy', {
            httpRequest: {},
            httpResponse: {},
            httpExchange: {},
          });
        }
      );
    });

    describe('within a timeout', function () {
      beforeEach(function () {
        handleRequest = function (req, res) {
          setTimeout(function () {
            res.setHeader('Content-Type', 'text/plain');
            res.end('foobar');
          }, 3);
        };
      });

      it('should not fail if its within the timeframe', function () {
        return expect(
          {
            url: serverUrl,
            timeout: 1000,
          },
          'to yield HTTP response satisfying',
          { body: 'foobar' }
        );
      });

      it('should fail if it is not within the timeframe', function () {
        return expect(
          expect(
            {
              url: serverUrl,
              timeout: 1,
            },
            'to yield HTTP response satisfying',
            'foobar'
          ),
          'when rejected to have message',
          `expected { url: 'http://${serverHostname}:${serverAddress.port}/', timeout: 1 }\n` +
            `to yield HTTP response satisfying 'foobar'\n` +
            `  expected a response within 1 ms`
        );
      });
    });

    describe('using a number as a shorthand for a response with that status code', function () {
      beforeEach(function () {
        handleRequest = function (req, res) {
          res.writeHead(412);
          res.end();
        };
      });

      it('should succeed', function () {
        return expect(
          `GET ${serverUrl}`,
          'to yield HTTP response satisfying',
          412
        );
      });

      it('should fail with a diff', function () {
        return expect(
          function () {
            return expect(
              `GET ${serverUrl}`,
              'to yield HTTP response satisfying',
              503
            );
          },
          'to error with',
          /HTTP\/1\.1 412 Precondition Failed \/\/ should be 503 Service Unavailable/
        );
      });
    });

    describe('with a JSON response', function () {
      beforeEach(function () {
        handleRequest = function (req, res) {
          res.setHeader('Content-Type', 'application/json');
          res.end('{"foo": 123}');
        };
      });

      it('should succeed', function () {
        return expect(`GET ${serverUrl}`, 'to yield HTTP response satisfying', {
          body: {
            foo: 123,
          },
        });
      });

      it('should fail with a diff', function () {
        return expect(
          expect(`GET ${serverUrl}`, 'to yield HTTP response satisfying', {
            body: {
              foo: 456,
            },
          }),
          'when rejected to have message',
          expect.it(function (message) {
            expect(
              message
                .replace(/^\s*Connection:.*\n/m, '')
                .replace(/\n\s*Content-Length:.*$|\s*Content-Length:.*\n/gm, '')
                .replace(
                  /\n\s*Transfer-Encoding:.*$|\s*Transfer-Encoding:.*\n/gm,
                  ''
                ),
              'to equal',
              `expected 'GET ${serverUrl}'\n` +
                `to yield HTTP response satisfying { body: { foo: 456 } }\n` +
                `\n` +
                `GET / HTTP/1.1\n` +
                `Host: ${serverHostname}:${serverAddress.port}\n` +
                `\n` +
                `HTTP/1.1 200 OK\n` +
                `Content-Type: application/json\n` +
                `\n` +
                `{\n` +
                `  foo: 123 // should equal 456\n` +
                `}`
            );
          })
        );
      });

      describe('with an expectation that requires async work', function () {
        it('should succeed', function () {
          return expect(
            `GET ${serverUrl}`,
            'to yield HTTP response satisfying',
            {
              body: {
                foo: expect.it('when delayed a little bit to equal', 123),
              },
            }
          );
        });

        it('should fail with a diff', function () {
          return expect(
            expect(`GET ${serverUrl}`, 'to yield HTTP response satisfying', {
              body: {
                foo: expect.it('when delayed a little bit to equal', 456),
              },
            }),
            'when rejected to have message',
            expect.it(function (message) {
              expect(
                message
                  .replace(/^\s*Connection:.*\n/m, '')
                  .replace(
                    /\n\s*Content-Length:.*$|\s*Content-Length:.*\n/gm,
                    ''
                  )
                  .replace(
                    /\n\s*Transfer-Encoding:.*$|\s*Transfer-Encoding:.*\n/gm,
                    ''
                  ),
                'to equal',
                `expected 'GET ${serverUrl}'\n` +
                  `to yield HTTP response satisfying { body: { foo: expect.it('when delayed a little bit to equal', 456) } }\n` +
                  `\n` +
                  `GET / HTTP/1.1\n` +
                  `Host: ${serverHostname}:${serverAddress.port}\n` +
                  `\n` +
                  `HTTP/1.1 200 OK\n` +
                  `Content-Type: application/json\n` +
                  `\n` +
                  `{\n` +
                  `  foo: 123 // expected: when delayed a little bit to equal 456\n` +
                  `}`
              );
            })
          );
        });
      });

      it('should send the correct Authorization header when specified in the headers object', function () {
        let authorizationHeader;
        handleRequest = function (req, res) {
          authorizationHeader = req.headers.authorization;
          res.end();
        };
        return expect(
          {
            url: `GET ${serverUrl}`,
            headers: {
              Authorization: 'foobar',
            },
          },
          'to yield HTTP response satisfying',
          200
        ).then(function () {
          expect(authorizationHeader, 'to equal', 'foobar');
        });
      });

      it('should send the correct Authorization header when the credentials are specified in the url', function () {
        let authorizationHeader;
        handleRequest = function (req, res) {
          authorizationHeader = req.headers.authorization;
          res.end();
        };
        return expect(
          {
            url: `GET http://foobar:quux@${serverHostname}:${serverAddress.port}/`,
          },
          'to yield HTTP response satisfying',
          200
        ).then(function () {
          expect(authorizationHeader, 'to equal', 'Basic Zm9vYmFyOnF1dXg=');
        });
      });
    });

    describe('with a request body stream', function () {
      beforeEach(function () {
        handleRequest = function (req, res) {
          req.pipe(res);
        };
      });

      it('should succeed', function () {
        const responseBodyStream = new stream.Readable();
        responseBodyStream._read = function (num, cb) {
          responseBodyStream._read = function () {};
          setTimeout(function () {
            responseBodyStream.push('foobar');
            responseBodyStream.push(null);
          }, 0);
        };

        return expect(
          {
            url: `PUT ${serverUrl}`,
            body: responseBodyStream,
          },
          'to yield HTTP response satisfying',
          {
            body: Buffer.from('foobar', 'utf-8'),
          }
        );
      });

      it('should fail if there was an error on the stream', function () {
        const erroringStream = new stream.Readable();
        erroringStream._read = function (num, cb) {
          setTimeout(function () {
            erroringStream.emit('error', new Error('Fake error'));
          }, 0);
        };

        return expect(
          {
            url: `PUT ${serverUrl}`,
            body: erroringStream,
          },
          'to yield HTTP response satisfying',
          new Error('Fake error')
        );
      });
    });
  });
});


test/browser/unexpectedHttp.js
describe('unexpectedHttp in the browser', function () {
  const expect = window.weknowhow.expect.clone();
  console.log('window', window.unexpectedHttp.default);
  expect.use(window.unexpectedHttp.default).use(window.magicPenPrism);

  it('should assert a content type for HTML', function () {
    return expect('GET /base/test/browser/index.html', 'to yield response', {
      headers: {
        'Content-Type': /^text\/html/,
      },
    });
  });

  it('should assert a content type for an image', function () {
    return expect('GET /base/test/browser/image.png', 'to yield response', {
      headers: {
        'Content-Type': /^image\/png/,
      },
    });
  });
});


karma.conf.js
const puppeteer = require('puppeteer');
process.env.CHROME_BIN = puppeteer.executablePath();

module.exports = function (config) {
  config.set({
    frameworks: ['mocha'],

    files: [
      {
        pattern: 'test/browser/index.html',
        watched: false,
        included: false,
        served: true,
        nocache: false,
      },
      {
        pattern: 'test/browser/image.png',
        watched: false,
        included: false,
        served: true,
        nocache: false,
      },
      './node_modules/unexpected/unexpected.js',
      './node_modules/magicpen-prism/magicPenPrism.min.js',
      './unexpectedHttp.min.js',
      './test/browser/unexpectedHttp.js',
    ],

    client: {
      mocha: {
        reporter: 'html',
        timeout: 60000,
      },
    },

    browsers: ['ChromeHeadless'],
  });
};


rollup.config.js
import nodeResolve from '@rollup/plugin-node-resolve';
import nodeGlobals from 'rollup-plugin-node-globals';
import json from '@rollup/plugin-json';
import commonjs from '@rollup/plugin-commonjs';
import nodePolyfills from 'rollup-plugin-polyfill-node';

const plugins = [
  commonjs(),
  nodePolyfills(),
  json(),
  nodeResolve({ preferBuiltins: false }),
  nodeGlobals(),
];

module.exports = [
  {
    input: 'lib/unexpectedHttp.js',
    output: {
      file: 'unexpectedHttp.js',
      name: 'unexpectedHttp',
      exports: 'named',
      format: 'umd',
      sourcemap: false,
      strict: false,
    },
    plugins,
  },
  {
    input: 'lib/unexpectedHttp.js',
    output: {
      file: 'unexpectedHttp.min.js',
      name: 'unexpectedHttp',
      exports: 'named',
      format: 'umd',
      sourcemap: false,
      strict: false,
    },
    plugins: [...plugins, require('rollup-plugin-terser').terser()],
  },
];


lib/unexpectedHttp.js
const http = require('http');
const https = require('https');
const messy = require('messy');
const _ = require('underscore');

module.exports = {
  name: 'unexpected-http',
  installInto: function (expect) {
    expect
      .child()
      .use(require('unexpected-messy'))
      .exportAssertion(
        [
          '<string|object> to yield [HTTP] response <any>',
          '<string|object> to yield [HTTP] response satisfying <any>',
        ],
        function (expect, subject, value) {
          const context = {};
          let expectedResponseError;
          if (
            Object.prototype.toString.call(value) === '[object Error]' ||
            value instanceof Error
          ) {
            expectedResponseError = value;
          } else if (
            value &&
            typeof value === 'object' &&
            (typeof Buffer === 'undefined' || !Buffer.isBuffer(value))
          ) {
            value = _.extend({}, value);
            if (value.error) {
              expectedResponseError = value.error;
            }
          }

          let request;
          return expect
            .promise(function (resolve, reject) {
              const requestOptions = {};
              const requestTimeout = subject.timeout || 0;
              const requestBody = subject.body;
              let requestBodyIsJson = false;
              let requestBodyIsStream = false;

              if (
                Array.isArray(requestBody) ||
                (requestBody &&
                  typeof requestBody === 'object' &&
                  (typeof Buffer === 'undefined' ||
                    !Buffer.isBuffer(requestBody)))
              ) {
                if (typeof requestBody.pipe === 'function') {
                  subject = _.omit(subject, 'body');
                  requestBodyIsStream = true;
                } else {
                  requestBodyIsJson = true;
                }
              }

              if (!subject.isMessyHttpRequest) {
                subject = new messy.HttpRequest(
                  subject && typeof subject === 'object'
                    ? _.omit(subject, 'timeout')
                    : subject
                );
              }

              const httpRequest = (context.httpRequest = subject);
              httpRequest.protocol = httpRequest.protocol || 'HTTP/1.1';
              if (requestBodyIsJson) {
                if (!httpRequest.headers.has('Content-Type')) {
                  httpRequest.headers.set('Content-Type', 'application/json');
                }
                httpRequest.body = JSON.stringify(httpRequest.body);
              }
              // Avoid using Transfer-Encoding: chunked for the request body as that confuses NGINX -- and we know the length
              // of the request body:
              if (
                !requestBodyIsStream &&
                !httpRequest.headers.has('Content-Length')
              ) {
                let contentLength = 0;
                if (httpRequest.unchunkedBody) {
                  contentLength = httpRequest.unchunkedBody.length;
                }
                httpRequest.headers.set('Content-Length', contentLength);
              }

              const headers = httpRequest.headers.toCanonicalObject();
              Object.keys(headers).forEach(function (headerName) {
                headers[headerName] = headers[headerName].join(', ');
              });
              _.extend(
                requestOptions,
                {
                  host:
                    httpRequest.host ||
                    (typeof window === 'object' &&
                    typeof window.location !== 'undefined'
                      ? window.location.hostname
                      : 'localhost'),
                  port:
                    httpRequest.port ||
                    (typeof window === 'object' &&
                      (typeof window.location !== 'undefined'
                        ? window.location.port
                        : window.location.protocol === 'https:'
                        ? 443
                        : 80)) ||
                    (httpRequest.encrypted ? 443 : 80),
                  method: httpRequest.method,
                  path: httpRequest.path + (httpRequest.search || ''),
                  headers,
                  mode: 'disable-fetch',
                },
                _.pick(subject, ['ca', 'cert', 'key', 'rejectUnauthorized'])
              );

              // set the protocol and scheme for HTTPS requests in the browser
              if (typeof window === 'object' && httpRequest.encrypted) {
                _.extend(requestOptions, {
                  scheme: 'https',
                  protocol: 'https:',
                });
              }

              let timeout = null;
              if (requestTimeout > 0) {
                timeout = setTimeout(function () {
                  expect.errorMode = 'nested';
                  try {
                    expect.fail(
                      'expected a response within {0} ms',
                      requestTimeout
                    );
                  } catch (err) {
                    reject(err);
                  }
                }, requestTimeout);
              }
              request = (httpRequest.encrypted ? https : http)
                .request(requestOptions)
                .on('response', function (response) {
                  const chunks = [];
                  response
                    .on('data', function (chunk) {
                      chunks.push(chunk);
                    })
                    .on('end', function () {
                      if (timeout) {
                        clearTimeout(timeout);
                      }

                      let body;
                      if (
                        chunks.length > 0 &&
                        typeof Buffer !== 'undefined' &&
                        Buffer.isBuffer(chunks[0])
                      ) {
                        body = Buffer.concat(chunks);
                      } else if (
                        chunks.length > 0 &&
                        chunks[0] instanceof Uint8Array
                      ) {
                        // Browserify: Make sure we end up with a Buffer instance:
                        if (chunks.length === 1) {
                          body = Buffer.from(chunks[0]);
                        } else {
                          let totalLength = 0;
                          let i;
                          for (i = 0; i < chunks.length; i += 1) {
                            totalLength += chunks[i].length;
                          }
                          body = Buffer.alloc(totalLength);
                          let offset = 0;
                          for (i = 0; i < chunks.length; i += 1) {
                            for (let j = 0; j < chunks[i].length; j += 1) {
                              body[offset] = chunks[i][j];
                              offset += 1;
                            }
                          }
                        }
                      } else {
                        // String or no chunks
                        body = chunks.join('');
                      }
                      const httpResponse = (context.httpResponse =
                        new messy.HttpResponse({
                          statusCode: response.statusCode,
                          statusMessage: http.STATUS_CODES[response.statusCode],
                          protocol: `HTTP/${response.httpVersionMajor || '1'}.${
                            response.httpVersionMinor || '1'
                          }`,
                          headers: response.headers,
                          unchunkedBody: body,
                        }));
                      resolve(
                        (context.httpExchange = new messy.HttpExchange({
                          request: httpRequest,
                          response: httpResponse,
                        }))
                      );
                    })
                    .on('error', reject);
                })
                .on('error', reject);

              if (request.xhr) {
                // In browserify: Make sure the response comes back as a Uint8Array
                request.xhr.responseType = 'arraybuffer';
              }

              if (requestBodyIsStream) {
                requestBody.pipe(request);
                requestBody.on('error', reject);
              } else {
                request.end(httpRequest.unchunkedBody);
              }
            })
            .then(
              function (httpExchange) {
                expect(expectedResponseError, 'to be undefined');
                return expect(httpExchange, 'to satisfy', { response: value });
              },
              function (err) {
                if (request) {
                  request.abort();
                }
                if (expectedResponseError) {
                  expect(err, 'to equal', expectedResponseError);
                } else if (err.code) {
                  // Socket or DNS error
                  expect.fail({
                    diff: function (output) {
                      output.error(err.message);
                      return output;
                    },
                    code: err.code,
                  });
                } else {
                  throw err;
                }
              }
            )
            .then(function () {
              return context;
            });
        }
      );
  },
};

module.exports.messy = messy;
